src/H5T.c

```diff
@@ -4294,12 +4730,28 @@ H5T__set_size(H5T_t *dt, size_t size)
     assert(!(H5T_ENUM == dt->shared->type && 0 == dt->shared->u.enumer.nmembs));
 
     if (dt->shared->parent) {
-        if (H5T__set_size(dt->shared->parent, size) < 0)
+        size_t new_size = size;
+
+        if (dt->shared->type == H5T_COMPLEX) {
+            if (!dt->shared->u.cplx.homogeneous)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_BADTYPE, FAIL,
+                            "heterogeneous complex number datatypes are currently unsupported");
+
+            if ((new_size % 2) != 0)
+                HGOTO_ERROR(H5E_DATATYPE, H5E_BADVALUE, FAIL,
+                            "new datatype size is not evenly divisible by 2");
+
+            new_size /= 2;
```

Comment:
The 'size' parameter is for the parent type (e.g. the way the array type works), so I don't think we want to divide by 2.

Reply:
When it comes to 'container' types, it's inconsistent whether setting the size is for the whole type or a base type, even though the docs for `H5Tset_size` say "H5Tset_size() sets the total size, size, in bytes, for a datatype". To follow the docs for complex numbers, I chose to have `H5Tset_size` set the size for the whole datatype, in which case the size of the parent type is half the datatype size.

Reply:
I could go either way on this, I suppose, since Complex straddles the line between array / compound types and atomic types.  I'd lean toward the way that arrays work, but don't feel strongly enough about it to insist.   @fortnern @derobins - where do you land on this?

Definitely needs to be documented, either way - could you add a blurb to the H5Tset_size()  documentation?

Reply:
See the changes to H5Tpublic.h

Reply:
> See the changes to H5Tpublic.h

OK, found that, thanks.   (I had been looking for them in H5Tmodule.h)

Still makes me feel like H5Tset_size() should be applied to the base type...

Reply:
I see that H5Tset_size is working on the entire type, but H5Tset_order and H5Tset_precision are working on the base type (which I agree with).  This just adds to the confusion here and I now think we should align all three and have H5Tset_size work on the base type also.

Reply:
The current situation makes perfect sense to me though because container types don't generally have an order or precision since they aren't atomic types.